### PR TITLE
possessions are getting better, but there is still a lot of improveme…

### PR DESCRIPTION
**Pull request by Jake0826**
…nt. I still need to look into making sure no one has possesion when ball is in the air

### CODE REVIEW

File: src/state.py [hunk 1]
```diff
@@ -383,7 +386,8 @@ def __init__(self, frameno: int) -> None:
         "frame number, Required: non-negative integer"
 
         # MUTABLE
-        self.players: dict[str, PlayerFrame] = {}  # ASSUMPTION: MULITPLE PEOPLE
+        # ASSUMPTION: MULITPLE PEOPLE
```
Comment (ericguo31) [inline src/state.py:389]: typo: should be multiple

File: src/state.py [hunk 2]
```diff
@@ -570,6 +574,13 @@ def __init__(self) -> None:
         self.team1: TeamStats = TeamStats()
         self.team2: TeamStats = TeamStats()
 
+        self.no_possession_distance = 100
+        self.player_scores = []  # to keep track of possession over frames
```
Comment (ericguo31) [inline src/state.py:578]: Define and distinguish each of player_scores, rolling_scores, dominant_possessions, and possession_history

File: src/state.py [hunk 3]
```diff
@@ -616,6 +627,16 @@ def populate_players_stats(self):
             for c in self.passes[p]:
                 self.players[p].passes.update({c: self.passes[p][c]})
 
+    def recompute_frame_count(self):
```
Comment (ericguo31) [inline src/state.py:630]: Change to compute_total_frames_per_player

File: src/state.py [hunk 4]
```diff
@@ -616,6 +627,16 @@ def populate_players_stats(self):
...
+        "recompute frame count of all players in frames"
```
Comment (ericguo31) [inline src/state.py:631]: Adjust documentation to reflect that this function computes the total number of frames each player_id appears in the video ==> used potentially to adjust for noise in reassigned ids

File: src/state.py [hunk 5]
```diff
@@ -735,138 +747,139 @@ def recompute_pass_from_possession(self):
...
+        # Initialize an empty list to store the dominant player for each frame
+        self.dominant_possessions = []
+
+        for i in range(len(self.frames)):
+            # Dictionary to sum up scores in the rolling window for each player
+            rolling_sum = {}
+
+            # Accumulate scores for each player in the rolling window of the last 50 frames
+            for j in range(max(0, i - 49), i + 1):
+                for player_id, score in self.rolling_scores[j].items():
```
Comment (Mikonooooo) [inline src/state.py:828]: This is pretty inefficient, usually when one computes a rolling windows, you continuously shift the window over by 1, adding the next frame's info while removing the last frame's info.  Can you make this more efficient. Another implementation you might consider is an exponential moving average

**Review by ericguo31 — CHANGES_REQUESTED** [review]
- Integrate all recompute_possessions_vx for x = 1, 2, 3, 4 into just one function recompute_possessions. There's no reason to separate it into four different functions here. The vx naming was misleading, as we thought it meant different versions of the same function when in reality they were all necessary components.
- Try to add more details on implementation in documentation and comments within code. Hard to interpret logic and would help future developers.
- See other comments on PR.

**Pull request closed without merging**